saltstack/base/salt/luks/bin/populate-luks-volume.sh

```diff
@@ -0,0 +1,168 @@
+#!/bin/bash
+
+set -ex -o pipefail
+
+export LUKS_VOLUME_NAME="cdp-luks"
+export MOUNT_POINT="/mnt/$LUKS_VOLUME_NAME"
+
+validate_absolute_path() {
+  local location="$1"
+  if [[ ! "$location" =~ ^/.* ]]
+  then
+    echo "Location \"$location\" must be an absolute path!"
+    exit 1
+  fi
+}
+
+validate_location_exists() {
+  local location="$1"
+  if [ ! -d "$location" ]
+  then
+    echo "Location \"$location\" must be an existing directory!"
+    exit 2
+  fi
+}
+
+create_location_if_necessary() {
+  local location="$1"
+  local owner=$2
+  local group=$3
+  local mode=$4
+  if [ ! -d "$location" ]
+  then
+    echo "Location \"$location\" does not exist; creating it with owner \"$owner\", group \"$group\", mode \"$mode\"."
+    install -o $owner -g $group -m $mode -d "$location"
+  fi
+}
+
+create_location_parent_tree_if_necessary() {
+  local source_location_parent
+  source_location_parent=$(dirname "$1")
+  if [ "$source_location_parent" != "/" ]
+  then
+    echo "Creating source location parent tree \"$source_location_parent\" if necessary."
+    source_location_parent="${source_location_parent#/}"
+    local source_location=""
+    local luks_location="$MOUNT_POINT"
+    local line
+    while read -r line
+    do
+      source_location="$source_location/$line"
+      rsync -dAogqk "$source_location" "$luks_location"
+      luks_location="$luks_location/$line"
+    done < <(echo "$source_location_parent" | awk -F / '{for (i=1; i<=NF; i++) print $i}')
+  fi
+}
+
+process_location() {
+  local source_location="$1"
+  echo "Processing source location \"$source_location\"."
+  validate_absolute_path "$source_location"
+  validate_location_exists "$source_location"
+  create_location_parent_tree_if_necessary "$source_location"
+  local luks_location="$MOUNT_POINT$source_location"
+  echo "Moving source location \"$source_location\" into LUKS location \"$luks_location\"."
+  local luks_location_parent
+  luks_location_parent=$(dirname "$luks_location")
+  mv "$source_location" "$luks_location_parent"
+  echo "Creating symlink \"$source_location\" pointing to \"$luks_location\"."
+  ln -s "$luks_location" "$source_location"
+}
+
+process_location_with_create() {
+  local location="$1"
+  validate_absolute_path "$location"
+  create_location_if_necessary "$@"
+  process_location "$location"
+}
+
+log_processing_needed() {
+  echo "Processing $1 locations."
+}
+
+log_processing_skipped() {
+  echo "No need to process $1 locations."
+}
+
+process_global_locations() {
+  log_processing_needed "global"
+  process_location_with_create "/etc/salt-bootstrap" root root 700
+  create_location_if_necessary "/etc/salt/pki" root root 744
+  process_location_with_create "/etc/salt/pki/minion" root root 700
```

Comment:
The explicit creation of `/etc/salt/pki` here is needed because it needs to have looser permissions than `/etc/salt/pki/minion`, if I understand correctly?

Reply:
That is correct. The permissions seen here are the exact ones currently found in our clusters (checked mostly on a green FreeIPA cluster, or sometimes on a green DL). I felt it more consistent to mirror the current behavior in these LUKS locations. These locations do not exist in the images, they only manifest upon the startup of the respective Salt processes. (I could not find any direct reference to these locations in CB repos, so I presume they are directly created by Salt Minion etc.)

On the other hand, I did not investigate further what the reason is for such discrepancies (like `/etc/salt/pki` vs `/etc/salt/pki/minion`); possibly these are created by separate processes having a different umask, or the same process explicitly using different access mode settings.